evm/src/cpu/shift.rs

```diff
@@ -0,0 +1,123 @@
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cpu::columns::CpuColumnsView;
+use crate::cpu::membus::NUM_GP_CHANNELS;
+use crate::memory::segments::Segment;
+
+pub(crate) fn eval_packed<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let is_shift = lv.op.shl + lv.op.shr;
+    let displacement = lv.mem_channels[1]; // holds the shift displacement d
+    let two_exp = lv.mem_channels[2]; // holds 2^d
+
+    // Not needed here; val is the input and we're verifying that output is
+    // val * 2^d (mod 2^256)
+    //let val = lv.mem_channels[0];
+    //let output = lv.mem_channels[NUM_GP_CHANNELS - 1];
+
+    let shift_table_segment = P::Scalar::from_canonical_u64(Segment::ShiftTable as u64);
+
+    let high_limbs_are_zero = lv.general.shift().displacement_high_limbs_are_zero;
+    yield_constr
+        .constraint(is_shift * (high_limbs_are_zero * high_limbs_are_zero - high_limbs_are_zero));
+
+    // Only lookup the shifting factor when displacement is < 2^32
+    yield_constr.constraint(is_shift * (two_exp.used - high_limbs_are_zero));
+    yield_constr.constraint(is_shift * (two_exp.is_read - high_limbs_are_zero));
+
+    let high_limbs_sum: P = displacement.value[1..].iter().copied().sum();
+    let high_limbs_sum_inv = lv.general.shift().high_limb_sum_inv;
+    // Verify that high_limbs_are_zero = 0 implies high_limbs_sum != 0 and
+    // high_limbs_are_zero = 1 implies high_limbs_sum = 0.
+    let t = high_limbs_sum * high_limbs_sum_inv - (P::ONES - high_limbs_are_zero);
+    yield_constr.constraint(is_shift * t);
+    yield_constr.constraint(is_shift * high_limbs_sum * high_limbs_are_zero);
+
+    // When the shift displacement is < 2^32, constrain the two_exp
+    // mem_channel to be the entry corresponding to `displacement` in
+    // the shift table lookup (will be zero if displacement >= 256).
+    let small_disp_filter = is_shift * high_limbs_are_zero;
+    yield_constr.constraint(small_disp_filter * two_exp.addr_context); // read from kernel memory
+    yield_constr.constraint(small_disp_filter * (two_exp.addr_segment - shift_table_segment));
+    yield_constr.constraint(small_disp_filter * (two_exp.addr_virtual - displacement.value[0]));
```

Comment:
Same as `two_exp.is_read`: this doesn’t need to be conditioned on `high_limbs_are_zero` because the value is ignored when `two_exp.used == P::ZEROS`.
```suggestion
    yield_constr.constraint(is_shift * two_exp.addr_context); // read from kernel memory
    yield_constr.constraint(is_shift * (two_exp.addr_segment - shift_table_segment));
    yield_constr.constraint(is_shift * (two_exp.addr_virtual - displacement.value[0]));
```